Mathlib/Tactic/Group.lean

```diff
@@ -0,0 +1,101 @@
+some `ring` invocations.
+
+## Tags
+
+group_theory
+-/
+
+namespace Mathlib.Tactic.Group
+
+open Lean
+open Lean.Meta
+open Lean.Parser.Tactic
+open Lean.Elab.Tactic
+
+-- The next three lemmas are not general purpose lemmas, they are intended for use only by
+-- the `group` tactic.
+@[to_additive]
+theorem zpow_trick {G : Type _} [Group G] (a b : G) (n m : ℤ) :
+    a * b ^ n * b ^ m = a * b ^ (n + m) := by rw [mul_assoc, ← zpow_add]
+#align tactic.group.zpow_trick Mathlib.Tactic.Group.zpow_trick
+#align tactic.group.zsmul_trick Mathlib.Tactic.Group.zsmul_trick
+
+@[to_additive]
+theorem zpow_trick_one {G : Type _} [Group G] (a b : G) (m : ℤ) :
+    a * b * b ^ m = a * b ^ (m + 1) := by rw [mul_assoc, mul_self_zpow]
+#align tactic.group.zpow_trick_one Mathlib.Tactic.Group.zpow_trick_one
+#align tactic.group.zsmul_trick_zero Mathlib.Tactic.Group.zsmul_trick_zero
+
+@[to_additive]
+theorem zpow_trick_one' {G : Type _} [Group G] (a b : G) (n : ℤ) :
+    a * b ^ n * b = a * b ^ (n + 1) := by rw [mul_assoc, mul_zpow_self]
+#align tactic.group.zpow_trick_one' Mathlib.Tactic.Group.zpow_trick_one'
+#align tactic.group.zsmul_trick_zero' Mathlib.Tactic.Group.zsmul_trick_zero'
+
+/-- Auxiliary tactic for the `group` tactic. Calls the simplifier only. -/
+syntax (name := aux_group₁) "aux_group₁" (ppSpace location)? : tactic
+
+macro_rules
+| `(tactic| aux_group₁ $[at $location]?) =>
+  `(tactic| simp (config := {decide := false}) only
+    [commutatorElement_def, mul_one, one_mul,
+      ←zpow_neg_one, ←zpow_ofNat, ←zpow_mul,
+      Int.ofNat_add, Int.ofNat_mul,
+      Int.mul_neg_eq_neg_mul_symm, Int.neg_mul_eq_neg_mul_symm, neg_neg,
+      one_zpow, zpow_zero, zpow_one, mul_zpow_neg_one,
+      ←mul_assoc,
+      ←zpow_add, ←zpow_add_one, ←zpow_one_add, zpow_trick, zpow_trick_one, zpow_trick_one',
+      tsub_self, sub_self, add_neg_self, neg_add_self]
+  $[at $location]?)
+
+/-- Auxiliary tactic for the `group` tactic. Calls `ring_nf` to normalize exponents. -/
+syntax (name := aux_group₂) "aux_group₂" (ppSpace location)? : tactic
+
+macro_rules
+| `(tactic| aux_group₂ $[at $location]?) =>
+  `(tactic| ring_nf $[at $location]?)
+
+/-- Tactic for normalizing expressions in multiplicative groups, without assuming
+commutativity, using only the group axioms without any information about which group
+is manipulated.
+
+(For additive commutative groups, use the `abel` tactic instead.)
+
+Example:
+```lean
+example {G : Type} [group G] (a b c d : G) (h : c = (a*b^2)*((b*b)⁻¹*a⁻¹)*d) : a*c*d⁻¹ = a :=
+begin
+  group at h, -- normalizes `h` which becomes `h : c = d`
+  rw h,       -- the goal is now `a*d*d⁻¹ = a`
+  group,      -- which then normalized and closed
+end
+```
+-/
+syntax (name := group) "group" (ppSpace location)? : tactic
+
+macro_rules
+| `(tactic| group $[at $location]?) =>
+  `(tactic| aux_group₁ $[at $location]? <;>
+            aux_group₂ $[at $location]? <;>
+            aux_group₁ $[at $location]?)
```

Comment:
It is probably the same as
```suggestion
  `(tactic| (aux_group₁ $[at $location]?;
            aux_group₂ $[at $location]?;
            aux_group₁ $[at $location]?))
```
since you don't create additional goals.

Reply:
It seems like the second and third tactics will sometimes complain that there are no goals to be solved.